index.js

```diff
@@ -1,47 +1,163 @@
 'use strict';
-const util = require('util');
-const fs = require('fs');
-const chalk = require('chalk');
+
 const Jimp = require('@sindresorhus/jimp');
-const termImg = require('term-img');
+const chalk = require('chalk');
+const {string: termImg} = require('term-img');
+const terminalCharWidth = require('terminal-char-width');
+
+const CHAR_HEIGHT = 16;
+const CHAR_WIDTH = CHAR_HEIGHT * terminalCharWidth;
+
+/**
+ * Check if two colors are equal
+ *
+ * Both has `alpha=0`, or `alpha>0` and equl all the other color values
+ *
+ * @returns {Boolean} colors are equal
+ */
+function equalColor({r, g, b, a}, {r: r2, g: g2, b: b2, a: a2}) {
+	if (a === 0 && a2 === 0) {
+		return true;
+	}
+
+	return a > 0 && a2 > 0 && r === r2 && g === g2 && b === b2;
+}
+
+function getBgColor({r, g, b, a}) {
+	// Full transparent
+	if (a === 0) {
+		return chalk.reset;
+	}
+
+	return chalk.bgRgb(r, g, b);
+}
+
+function getFgColor({r, g, b}) {
+	return chalk.rgb(r, g, b);
+}
+
+function getGlyph(upper, lower) {
+	if (upper.a) {
+		if (lower && equalColor(upper, lower)) {
+			return '█';
+		}
+
+		return '▀';
+	}
+
+	if (lower && lower.a) {
+		return '▄';
+	}
+
+	return ' ';
+}
+
+async function render(fileBuffer, {asArray, height, width}) {
+	const image = await Jimp.read(fileBuffer);
+	const {bitmap} = image;
 
-const PIXEL = '\u2584';
-const readFile = util.promisify(fs.readFile);
+	if (height === undefined && width === undefined) {
+		height = bitmap.height / CHAR_HEIGHT;
+		width = bitmap.width / CHAR_WIDTH;
+	} else if (height === undefined) {
+		height = bitmap.height * terminalCharWidth * width / bitmap.width;
+	} else {
+		width = bitmap.width / terminalCharWidth * height / bitmap.height;
+	}
+
+	// Each character has two vertical blocks, so we double the number of rows
+	height *= 2;
 
-async function render(buffer) {
-	const image = await Jimp.read(buffer);
-	const columns = process.stdout.columns || 80;
```

Comment:
The image should never be wider than the terminal. That makes it just a mess of colors.

<img width="941" alt="screen shot 2018-05-16 at 23 39 27" src="https://user-images.githubusercontent.com/170270/40130838-642fc830-5962-11e8-9777-3b3861f2e1b6.png">

Reply:
By default, it should work like before and just fill the terminal while preserving aspect ratio.

Reply:
Ok :+1: